src/main/java/org/broadinstitute/hellbender/tools/spark/ParallelCopyGCSDirectoryIntoHDFSSpark.java

```diff
@@ -0,0 +1,171 @@
+                gcsPaths = Files.list(inputGCSNIOPath).collect(Collectors.toList());
+            } else {
+                logger.warn("single file");
+                gcsPaths = Collections.singletonList(inputGCSNIOPath);
+            }
+
+            List<Tuple2<String, Integer>> chunkList = new ArrayList<>();
+            for (Path path : gcsPaths) {
+
+                if (Files.isDirectory(path)) {
+                    logger.info("skipping directory " + path);
+                    continue;
+                }
+
+                final long size = Files.size(path);
+                final long chunks = size / chunkSize + 1;
+                logger.info("processing path " + path + ", size = " + size + ", chunks = " + chunks);
+
+                for (int i = 0; i < chunks; i++) {
+                    chunkList.add(new Tuple2<>(path.toUri().toString(), i));
+                }
+            }
+
+            if (chunkList.size() == 0) {
+                logger.info("no files found to copy");
+                return;
+            }
+
+            final JavaPairRDD<String, Integer> chunkRDD = ctx.parallelizePairs(chunkList, chunkList.size());
+
+            final JavaPairRDD<String, Tuple2<Integer, String>> chunkMappingRDD =
+                    chunkRDD.mapToPair(p -> new Tuple2<>(p._1(), readChunkToHdfs(auth, p._1(), chunkSize, p._2(), outputDirectoryFinal)));
+
+            final Map<String, Iterable<Tuple2<Integer, String>>> chunksByFilePath = chunkMappingRDD.groupByKey().collectAsMap();
+
+            for (Path path : gcsPaths) {
+
+                if (Files.isDirectory(path)) {
+                    continue;
+                }
+
+                final String filePath = path.toUri().toString();
+                final Iterable<Tuple2<Integer, String>> chunkListForFile = chunksByFilePath.get(filePath);
+                final String basename = path.getName(path.getNameCount() - 1).toString();
+                final org.apache.hadoop.fs.Path outFilePath = new org.apache.hadoop.fs.Path(outputDirectoryFinal + "/" + basename);
+                fs.createNewFile(outFilePath);
+
+                SortedMap<Integer, String> chunkMap = new TreeMap<>();
+                for (Tuple2<Integer, String> entry : chunkListForFile) {
+                    chunkMap.put(entry._1(), entry._2());
+                }
+
+                org.apache.hadoop.fs.Path[] chunkPaths = new org.apache.hadoop.fs.Path[chunkMap.size()];
+
+                final Iterator<Integer> iterator = chunkMap.keySet().iterator();
+                while (iterator.hasNext()) {
+                    final Integer next =  iterator.next();
+                    final String chunkPath = chunkMap.get(next);
+                    chunkPaths[next] = new org.apache.hadoop.fs.Path(chunkPath);
+                }
+
+                fs.concat(outFilePath, chunkPaths);
+            }
+
+
+
+        } catch (IOException e) {
+            throw new GATKException(e.getMessage(), e);
```

Comment:
I added an extra catch for when the GCS input path can't be found, with a reminder to use a trailing slash if you want to copy a directory (which is how I seem to mess this up a lot).